git-bug-hash

makrushin-evgenii opened the pull request:
Возникло пару вопросов при реализации:

1. Мы выводим тот же хэш, который подали на вход git bug, или хэш дерева?
2. Нужно ли выводить поле status?

makrushin-evgenii committed:
Create commands

makrushin-evgenii committed:
Delete commands

makrushin-evgenii committed:
Create git-bug-hash

Вывод: hash, summary, description.

teabagk7 commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+	echo "bug $HASH"
+
+	# STATUS=$( git cat-file -p $STATUS_HASH 0>&1)
+	# echo "Status: $STATUS"
```
Status нужен, это же такой же атрибут описанный в README

dimonchik0036 replied:
Done

makrushin-evgenii commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Реализовать команду просмотра тикета ('git bug <hash>')](https://trello.com/c/4i1kE5FY/4-git-bug-hash)

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+
+get_tree_hash() 
+{
+	touch temp$$
```
1) А зачем touch? Файл должен создаться, когда начнём в него писать через `>`
2) Предлагаю имя `temp$$` вынести в отдельную глобальную константу, а не повторять по всему скрипту.

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+	
+	read line < temp$$
+	set $line
+	TREE_HASH=$(echo "$2")
```
если передаёшь наверх значение через глобальную переменную, то лучше тогда её объявить в начале скрипта.

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+
+	get_tree_hash "$1"
+
+
```
Странное форматирование - двойной отступ. Может, все три команды рядом положить? Вполне будет гармонично смотреться.

dimonchik0036 replied:
Done

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+	get_tree_hash "$1"
+
+
+	git cat-file -p $TREE_HASH 1> temp$$
```
а зачем тут явно stdout поток как `1` указывать? просто `>` вполне бы хватило

dimonchik0036 replied:
Done

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+	
+	rm -f temp$$
+
+
```
снова двойной отступ

dimonchik0036 replied:
Done

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+	rm -f temp$$
+
+
+	echo "bug $HASH"
```
может, _issue_, а не _bug_?

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+
+git_bug_hash()
+{
+	HASH="$1"
```
что будет, если входного параметра не передано? Не вижу ни одной проверки.

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+	# STATUS=$( git cat-file -p $STATUS_HASH 0>&1)
+	# echo "Status: $STATUS"
+
+	SUMMARY=$( git cat-file -p $SUMMARY_HASH 0>&1)
```
И для статуса, и для summary тот же функционал уже реализован в git-bug (но по-другому).
Во-первых, где более правильная реализация? Во-вторых, предлагаю не копипастить, а создать отдельный общий скрипт с названием вроде `git-bug-common.sh` (@AnnaKhismatullina подскажет, возможно, более правильное название) и в нём определить общие функции вроде _get_status()_ и _get_summary()_.
Включать этот скрипт себе через `source` или через `.` и дёргать эти общие функи из всех мест.

kevromster commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+	SUMMARY=$( git cat-file -p $SUMMARY_HASH 0>&1)
+	echo "Summary: $SUMMARY"
+		
+	DESCRIPTION=$(git cat-file -p $DESCRIPTION_HASH 0>&1)
```
А что здесь `0` означает?

teabagk7 commented on src/commands/git-bug-hash:
```diff
@@ -0,0 +1,58 @@
+
+get_tree_hash() 
+{
+	touch temp$$
```
У меня предложение дописать скрипт git bug.
Ведь по смыслу это относится к нему.
Еще возникает вопрос зачем нам выводить хэш задачи, если он уже ввёл этот хэш? Может отказаться на выводе от него?
Скрипт не выдаст актуальную версию, т.к хэш - идентификатор _issue_, по которому мы должны выдавать актуальную версию. Сейчас выдается самая первая версия issue как понимаю.

kevromster replied:
поддерживаю идею дописать _git-bug_ вместо отдельного скрипта.

dimonchik0036 committed:
Update git-bug-hash

dimonchik0036 commented:
Постарался учесть все предыдущие замечания. Правильно ли объявлен `TREE_HASH` в начале?

mr1name reviewed (changes requested):
Что за ошибка "This hash ($2) is not found."? Сообщение, что этот скрипт не нашел какой-то хэш может означать в глазах пользователя что угодно. Непонятно является ли это какой-то внутренней ошибкой, или означает ли это, что ты просто ввел что-то не то. Человек, который это увидит, скорее всего пойдет искать в интернет, что это означает, хотя скрипт должен был сообщить одну простую вещь - issue с таким идентификатором не существует. А то, что нет какого-то хэша (от которых мы возможно даже откажемся и будем как идентификатор использовать что-то другое) только запутает человека и отнимет у него лишнее время.

mr1name reviewed (changes requested):
И еще по сообщениям, что такое "Please enter the correct data."? Откуда я знаю что является корректным? Если пользователь неправильно ввел команду, то ему сразу и надо показать, как её использовать правильно, так, чтобы он понял, что он сделал не так. Ваши сообщения об ошибках по информативности равносильны завершению программы с кодом -1, поэтому прошу присмотреться к ним повнимательнее.

mr1name reviewed (dismissed):
Информативность работы самой команды тоже близка к нулю. Как мне узнать кто создал Issue и когда, как понять, когда изменялся его статус и кто его изменял, как менялось его описание и остальные атрибуты? Плюс одного линейного пути тут недостаточно. Есть многие вещи, которые хочется видеть не всегда, например, полный список всех изменений статуса Issue с датами и именами, то есть нужны еще дополнительные параметры, которые позволяли бы видеть доп. информацию. Это самая центральная команда в проекте, потому, что если у пользователя просто не будет возможность получить большую часть какой-то полезной информации (которая у нас есть!), то никого смысла в самом проекте и нет. Для той информации которую вы сейчас выдаете пользователю, нам достаточно было бы просто хранить текстовый файл с issue и изменять его когда надо, потому что ни а какой истории изменений, и вообще всех возможностей, которые мы получили благодаря гиту, тут речи даже не идет.

dimonchik0036 committed:
Update git-bug

dimonchik0036 committed:
Delete git-bug-hash

Commit:
Merge branch 'master' into feature/4

Commit:
4 Fixed whitespaces in git-bug

akhismat commented on src/git-bug:
```diff
@@ -139,13 +139,13 @@ main()
-		exit 1
-	fi
-	"$workingdir/git-bug-$subcommand" "$@"
+    local workingdir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
```
Посколько не все команды будут содержаться в этом файле, а только git bug и git bug hash, то эта проверка нам понадобится, так что я ее восстановила из мастера.

akhismat commented on src/git-bug:
```diff
@@ -1,99 +1,151 @@
+            list_issues "O"
+            exit 0
+            ;;
+        *)
```
Из-за этого условия, все команды git bug * будут обрабатываться git_bug_hash, что не верно.
Например, попытка сделать git bug new:
_bash -x src/git-bug new_                                                                                                                                                                                                                                         
```
+ main new
+ REF=refs/heads/ISSUES
+ '[' 1 -lt 1 ']'
+ local subcommand=new
+ shift
+ case "$subcommand" in
+ git_bug_hash new
```
_usage: git bug [-h | -a | --open | --resolved | <subcommand>]
    <subcommand> can be :
    new             create new issue
    resolve         resolve issue
    hash            prints issue information_

akhismat replied:
Чтобы это исправить нужно запились более аккуратный regexp, чем просто *.

dimonchik0036 replied:
Done

makrushin-evgenii committed:
FIX git-bug-hash call

dimonchik0036 committed:
Update git-bug-hash out

dimonchik0036 committed:
Update git-bug

akhismat commented on src/git-bug:
```diff
@@ -1,99 +1,154 @@
+    esac
+    
+    # subcommand - hex число => git-bug-hash 
+    if expr "$subcommand" : "[a-f0-9]*" 1>/dev/null 
```
Не обязательно выносить проверку в отдельный if, лучше было заменить regex в case вместо _*_ на _[a-f0-9]*_
типа:
```
        [a-f0-9]*)
            git_bug_hash "$subcommand"
            exit 0
            ;;
```

makrushin-evgenii replied:
Возможно bug hash будет логичнее вызывать там же, где и другие команды, а в case оставить обработку флагов.

kevromster replied:
Я бы оставил в отдельном `if`. Но:
1) тут забыто A-F
2) текущий regex, кажется, пройдёт любое слово на _a_. А если у нас такая команда появится в будущем?

akhismat replied:
1. Честно не помню, в hash'е git'a заглявные буквы
2. Попробовала 
```
git bug add
git bug abcd
```
regex не сработал)

kevromster replied:
шестнадцатеричные циферки я могу и заглавными захотеть написать (или какой-нибудь инструмент мне их так отдаст).
Кстати, add - это валидное начало хэша.

dimonchik0036 replied:
Done

dimonchik0036 committed:
FIX commands

dimonchik0036 committed:
FIX git-bug-hash

dimonchik0036 committed:
Update short-hash

Tatiotaeryy commented on src/git-bug:
```diff
@@ -1,99 +1,148 @@
-				;;
-		esac
+    local issue="$1"
+    local status_hash="$(git cat-file -p "$(git cat-file -p $issue | awk ' /tree/ {print $2}')" | awk ' /status/ {print $3}')"
```
Не бейте меня палками, но в этом коде встречаются два различных подхода которые могу заставить пользователя попотеть:
1. В качестве hash команда "git bug hash" принимает хэш актуального коммита issue (последнего).
2. В качестве hash команда "git bug" возвращает хэш корневого коммита issue (первого), которое если верить модели выступает в качестве "имени" issue.

Это не помешает работе скрипта, но вызовет негодование пользователя, который попробовав посмотреть issue по хэшу выведенному git bug увидит, что issue не найдено, но при этом сможет его редактировать)

dimonchik0036 replied:
Не совсем понятен комментарий. При поиске хэша, выведенного через `git bug`, корректно отображается полная информация по нему.

Tatiotaeryy replied:
Это правда, не совсем понятно написал)
Поиск по хэшу выведенному через git bug будет выдавать всегда одну и ту же информацию.

makrushin-evgenii committed:
rec

Tatiotaeryy committed:
All what you wanted

makrushin-evgenii committed:
fix output

kevromster reviewed (approved):
LGTM

makrushin-evgenii committed:
author and date last

makrushin-evgenii committed:
HOT FIX

Fix #1: выводим дату создания, а не последнюю
Fix #2: выводим автора первого коммита
